### PR TITLE
use shell struct in shell-local provisioner and post-processor.

### LINKED ISSUE #8338: Implement common/shell embedded struct in shell-local provisioner and post-processor
We have an internal "shell" struct that is used for shell, powershell, and windows-shell provisioners. It makes sense to implement this within shell-local to reduce duplicate configuration keys and enhance test coverage.

### PR DESCRIPTION
**Pull request by SwampDragons**
This refactors the code to use the common shell struct in the shell-local provisioner and post-processor, deduplicating several options. 

It required splitting the shell.Provisioner struct into two components, one that contained remote-specific options.

It also required adding the valid_exit_codes options to the shell-local provisioner and post-processors. 

I've noticed that we aren't auto-generating partials for the common shell struct; they're still all hardcoded. That should probably be done as a follow-up to this PR, but it felt out of scope here. 

Closes #8338

### CODE REVIEW

File: common/shell-local/config.go [hunk 1]
```diff
@@ -19,14 +20,17 @@ import (
 type Config struct {
 	common.PackerConfig `mapstructure:",squash"`
 
+	shell.Provisioner `mapstructure:",squash"`
+
 	// ** DEPRECATED: USE INLINE INSTEAD **
 	// ** Only Present for backwards compatibility **
 	// Command is the command to execute
 	Command string
```
Comment (azr) [inline common/shell-local/config.go]: [This should probably be another PR?] Should we get rid of Command here with a fixer; since it is deprecated ? 🤔

File: common/shell-local/run.go [hunk 2]
```diff
@@ -97,12 +97,9 @@ func Run(ctx context.Context, ui packer.Ui, config *Config) (bool, error) {
...
 				script)
 		}
-		if cmd.ExitStatus() != 0 {
-			return false, fmt.Errorf(
-				"Erroneous exit code %d while executing script: %s\n\n"+
-					"Please see output above for more information.",
-				cmd.ExitStatus(),
-				script)
+
+		if err := config.ValidExitCode(cmd.ExitStatus()); err != nil {
```
Comment (azr) [inline common/shell-local/run.go]: nice !

File: provisioner/powershell/provisioner.go [hunk 3]
```diff
@@ -40,6 +40,8 @@ var psEscape = strings.NewReplacer(
 type Config struct {
 	shell.Provisioner `mapstructure:",squash"`
 
+	shell.ProvisionerRemoteSpecific `mapstructure:",squash"`
```
Comment (nywilken) [inline provisioner/powershell/provisioner.go]: Nice :100:

File: common/shell-local/config.go [hunk 4]
```diff
@@ -19,14 +20,17 @@ import (
 type Config struct {
 	common.PackerConfig `mapstructure:",squash"`
 
+	shell.Provisioner `mapstructure:",squash"`
```
Comment (nywilken) [inline common/shell-local/config.go]: I believe this also embeds `common.PackerConfig` so we might be able to get rid of line 21

Comment (SwampDragons) [reply common/shell-local/config.go]: Ah, I'll check.

Comment (SwampDragons) [reply common/shell-local/config.go]: yep, you're right.

File: common/shell-local/config.go (hunk 1 above)
Comment (SwampDragons) [reply common/shell-local/config.go]: Yeah, we could probably ditch it now; it's been over a year since I made the change.
